test/lit/passes/stack-ir-non-nullable.wast

```diff
@@ -0,0 +1,836 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+;; RUN: wasm-opt %s --generate-stack-ir --optimize-stack-ir --shrink-level=1 \
+;; RUN:   -all --print-stack-ir | filecheck %s
+
+;; Shrink level is set to 1 to enable local2stack in StackIR opts.
+
+(module
+ ;; CHECK:      (func $if (type $1) (param $param (ref eq)) (result (ref eq))
+ ;; CHECK-NEXT:  (local $temp (ref eq))
+ ;; CHECK-NEXT:  local.get $param
+ ;; CHECK-NEXT:  local.set $temp
+ ;; CHECK-NEXT:  local.get $temp
+ ;; CHECK-NEXT:  i32.const 0
+ ;; CHECK-NEXT:  i31.new
+ ;; CHECK-NEXT:  ref.eq
+ ;; CHECK-NEXT:  if
+ ;; CHECK-NEXT:   i32.const 1
+ ;; CHECK-NEXT:   i31.new
+ ;; CHECK-NEXT:   local.set $temp
+ ;; CHECK-NEXT:  else
+ ;; CHECK-NEXT:   i32.const 2
+ ;; CHECK-NEXT:   i31.new
+ ;; CHECK-NEXT:   local.set $temp
+ ;; CHECK-NEXT:  end
+ ;; CHECK-NEXT:  local.get $temp
+ ;; CHECK-NEXT: )
+ (func $if (param $param (ref eq)) (result (ref eq))
+  (local $temp (ref eq))
+  ;; Copy the param into $temp. $temp is then set in both arms of the if, so
+  ;; it is set before the get at the end of the function, but we still need to
+  ;; keep this set for validation purposes. Specifically, there is a set of
+  ;; $temp followed by a get of it in the if condition, which local2stack could
+  ;; remove in principle, if not for that final get at the function end.
+  (local.set $temp
+   (local.get $param)
+  )
+  (if
+   (ref.eq
+    (local.get $temp)
+    (i31.new
+     (i32.const 0)
+    )
+   )
+   (local.set $temp
+    (i31.new
+     (i32.const 1)
+    )
+   )
+   (local.set $temp
+    (i31.new
+     (i32.const 2)
+    )
+   )
+  )
+  (local.get $temp)
+ )
+
+ ;; CHECK:      (func $if-no-last-get (type $1) (param $param (ref eq)) (result (ref eq))
+ ;; CHECK-NEXT:  (local $temp (ref eq))
+ ;; CHECK-NEXT:  local.get $param
+ ;; CHECK-NEXT:  i32.const 0
+ ;; CHECK-NEXT:  i31.new
+ ;; CHECK-NEXT:  ref.eq
+ ;; CHECK-NEXT:  if
+ ;; CHECK-NEXT:   i32.const 1
+ ;; CHECK-NEXT:   i31.new
+ ;; CHECK-NEXT:   local.set $temp
```

Comment:
We don't optimize out these dead sets?

Reply:
Other passes would remove those. We don't try to in this pass.